solvers/mosek/mosekbackend.cc

```diff
@@ -378,135 +378,174 @@ double MosekBackend::MIPGapAbs() {
+        else skx[j] = MSK_SK_SUPBAS;
+
+        break;
+      default:
+        MP_RAISE(fmt::format("Unknown AMPL var status value: {}", stt[j]));
     }
   }
-  MOSEK_SetBasis(lp(), stt.data(), NULL);
-  */
+
+  // Input the variable status keys
+  MOSEK_CCALL(MSK_putskx(lp(), MSK_SOL_BAS, (MSKstakeye *)skx));
 }
 
 void MosekBackend::ConStatii(ArrayRef<int> cst) {
-  // TODO
-  /*
   std::vector<int> stt(cst.data(), cst.data() + cst.size());
-  for (auto& s : stt) {
-    switch ((BasicStatus)s) {
-    case BasicStatus::bas:
-      s = MOSEK_BASIS_BASIC;
-      break;
-    case BasicStatus::none:   // for 'none', which is the status
-    case BasicStatus::upp:    // assigned to new rows, it seems good to guess
-    case BasicStatus::sup:    // a valid status.
-    case BasicStatus::low:    // 
-    case BasicStatus::equ:    // For active constraints, it is usually 'sup'.
-    case BasicStatus::btw:    // We could compute slack to decide though.
-      s = MOSEK_BASIS_SUPERBASIC;
-      break;
-    default:
-      MP_RAISE(fmt::format("Unknown AMPL con status value: {}", s));
+  int *skc = (int *) MSK_calloctask(lp(), stt.size(), sizeof(int));
+
+  // TODO: remove this if we are sure that it is the same.
+  MSKint32t numcon;
+  MOSEK_CCALL(MSK_getnumcon(lp(), &numcon));
+  assert(numcon == cst.size());
+
+  // The status key of a constraint is the status key of the logical (slack) variable assigned to it.
+  // The slack is defined as: l <= a'x <= u rewritten as a'x - s = 0, l <= s <= u.
+  for (auto j = 0; j < stt.size(); j++)
+  {
+    skc[j] = MSK_SK_UNK;
+    switch ((BasicStatus)stt[j])
+    {
+      case BasicStatus::bas:
+        skc[j] = MSK_SK_BAS;
+        break;
+      case BasicStatus::sup:
+      case BasicStatus::btw:
+        skc[j] = MSK_SK_SUPBAS;
+        break;
+      case BasicStatus::low:
+        skc[j] = MSK_SK_LOW;
+        break;
+      case BasicStatus::upp:
+        skc[j] = MSK_SK_UPR;
+        break;
+      case BasicStatus::equ:
+        skc[j] = MSK_SK_FIX;
+        break;
+      case BasicStatus::none:
+        /// 'none' is assigned to new constraints (added after solving).
+        /// Compute low/upp/fix/sup, depending on where initial value 0.0 is between the bounds
+        MSKboundkeye bk;
+        MSKrealt bl, bu, tolx;
+        MOSEK_CCALL(MSK_getconbound(lp(), (MSKint32t)j, &bk, &bl, &bu));
+        // MSK_DPAR_DATA_TOL_X is both for variables and constraints
+        MOSEK_CCALL(MSK_getdouparam(lp(), MSK_DPAR_DATA_TOL_X, &tolx));
+
+        if (0.0 < bl + tolx) skc[j] = MSK_SK_LOW;
+        else if (bu - tolx < 0.0) skc[j] = MSK_SK_UPR;
+        else if (bu - bl < tolx) skc[j] = MSK_SK_FIX;
+        else skc[j] = MSK_SK_SUPBAS;
+
+        break;
+      default:
+        MP_RAISE(fmt::format("Unknown AMPL var status value: {}", stt[j]));
     }
   }
-  MOSEK_SetBasis(lp(), NULL, stt.data());
-  */
+
+  // Input the variable status keys
+  MOSEK_CCALL(MSK_putskc(lp(), MSK_SOL_BAS, (MSKstakeye *)skc));
```

Comment:
Same here. Vector stt is allocated before, so why skc? We should avoid calloc() and Co, STL containers do the job cleaner.

Reply:
Same here, using vector.